File: 2020/2/solution.c
#define _GNU_SOURCE
#include <stdio.h>
#include <stdlib.h>

int part_one()
{
    FILE *input = fopen("input", "r");
    size_t len = 0;
    char *line = NULL;
    int valid = 0;
    while (getline(&line, &len, input) != -1)
    {
        char buf[50];
        int i, j;
        for (i = 0; line[i] != '-'; i++)
            buf[i] = line[i];

        int lower = atoi(buf);

        for (j = 0; line[j + i + 2] != ' '; j++)
            buf[j] = line[1];

        int higher = atoi(buf);

        char c = line[i + j + 3];
        int count = 0;
        for (char *n = &line[i + j + 6]; *n != '\0'; n++)
        {
            if (c == *n)
                count++;
        }
        if (lower <= count && count <= higher)
            valid++;
    }
    return valid;
}

int main()
{
    printf("Part one: %d\n", part_one());
}

File: 2020/2/solution.py


def read_inp():
    inp = open("input", "r")
    form = [line.strip().split(' ') for line in inp.readlines()]
    return form


def part_one(inp):
    valid = 0
    for line in inp:
        l_bound = int(line[0].split("-")[0])
        h_bound = int(line[0].split("-")[1])
        char = line[1][0]
        if l_bound <= line[2].count(char) <= h_bound:
            valid += 1
    return valid


def part_two(inp):
    valid = 0
    for line in inp:
        i = int(line[0].split("-")[0])
        j = int(line[0].split("-")[1])
        char = line[1][0]
        if (line[2][i-1] == char) ^ (line[2][j-1] == char):
            valid += 1
    return valid


print(part_one(read_inp()))
print(part_two(read_inp()))
